url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/PartsRobotAgent.java

```diff
@@ -0,0 +1,531 @@
+						if (arm.part.type.getName() != "Dummy") {
+							PlacePart(arm);
+							return true;
+						}
+					}
+				}
+			}
+
+			if (allArmsEmpty()) {
+				status = PartsRobotStatus.PICKING_UP;
+				return true;
+			}
+		}
+
+		// Checks if there is an empty arm, if there is it fills it with a
+		// good part that the kit needs
+
+		if (IsAnyArmEmpty()) {
+			status = PartsRobotStatus.PICKING_UP;
+			List<Nest> readyNests = new ArrayList<Nest>(GoodParts.keySet());
+			// synchronized (GoodParts) {
+			for (Nest nest : readyNests) {
+				// A nest contains a single part type
+				Part part = GoodParts.get(nest).get(0);
+				synchronized (MyKits) {
+					// print("Size of MyKits: " +
+					// MyKits.size());
+					for (MyKit mk : MyKits) {
+						// Checking if the good part is needed by
+						// either kit
+						// print("Kit needs: " +
+						// mk.kit.partsExpected.getConfig().toString());
+						if (NumTotalPartsNeeded(part) > NumPartsInHand(part)) {
+							print("Found a part I need of type " + part.type.getName() + " for kit "
+									+ MyKits.indexOf(mk) + " " + mk.kit.PartsStillNeeded());
+							synchronized (Arms) {
+								for (Arm arm : Arms) {
+									if (arm.AS == ArmStatus.EMPTY) {
+										// Find the empty arm
+										PickUpPart(arm, part, nest);
+										return true;
+									}
+								}
+							}
+
+						} // Why is this so awful
+					}
+				}
+			}
+			// }
+		} else {
+			status = PartsRobotStatus.PLACING;
+		}
+
+		if (System.currentTimeMillis() - time.getTime() > 5000) {
+			// Last rule is to place parts if the parts robot has been idle
+			// too long
+			time.setTime(System.currentTimeMillis());
+			status = PartsRobotStatus.PLACING;
+			return true;
+		}
+
+		timer.schedule(new TimerTask() {
+			// hack to force the partsrobot to attempt to place parts
+			// sleep. Fires after 5.001 seconds.
+			@Override
+			public void run() {
+				stateChanged();
+			}
+		}, 5001);
+
+		/*
+		 * Tried all rules and found no actions to fire. Return false to the main loop of abstract base class Agent and
+		 * wait.
+		 */
+		return false;
+	}
+
+	/*
+	 * Actions
+	 */
+
+	private void PickUpPart(Arm arm, Part part, Nest nest) {
+		time.setTime(System.currentTimeMillis());
+		print("Picking up part" + getPartTypesInArms());
+		// synchronized (Arms) {
+
+		GoodParts.remove(nest);
+		arm.AS = ArmStatus.FULL;
+		arm.part = part;
+		// Tells the graphics to pickup the part
+		if (partsRobotGraphics != null) {
+			partsRobotGraphics.pickUpPart(part.partGraphics, Arms.indexOf(arm));
+			try {
+				// print("Blocking");
+				animation.acquire();
+			} catch (InterruptedException e) {
+				e.printStackTrace();
+			}
+			// print("Got permit");
+		}
+
+		// Only takes 1 part from a nest at a time
+		nest.msgTakingPart(part);
+		nest.msgDoneTakingParts();
+
+		stateChanged();
+		// }
+	}
+
+	private void PlacePart(Arm arm) {
+
+		print("Placing part");
+		synchronized (MyKits) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 The code acquires two synchronization objects in **two different orders**, which can cause **circular wait** (deadlock).

 The **first order** is: first acquire `Arms` and then `MyKits`:

- enter `pickAndExecuteAnAction()` at** line 185**

- **acquire synchronization** on `Arms` at** line 201**

- call `PlacePart()` at** line 206**

- **acquire synchronization** on `MyKits` at** line 318**



 The **second order** is: first acquire `MyKits` and then `Arms`:

- enter `pickAndExecuteAnAction()` at** line 185**

- **acquire synchronization** on `MyKits` at** line 229**

- **acquire synchronization** on `Arms` at** line 240**



**Fix**  
 The fix is to acquire synchronization in only one order. Typically you can do this by adding extra synchronization to the synchronization order you want to prohibit.